.circleci/config.yml

```diff
@@ -37,10 +72,147 @@ jobs:
           name: 'Ensure clean dist diff'
           command: make validate-no-changes
 
+  build-docs:
+    executor: default_build_environment
+    working_directory: ~/repo
+    
+    steps:
+      - checkout
+      - base-tools/install
+      - bootstrap_node
+      - bootstrap_pipenv
+      - run:
+          name: 'Typedoc'
+          command: make docs
+      - run:
+          name: "MkDocs"
+          command: make build-mkdocs
+      - persist_to_workspace:
+          root: ~/repo
+          paths:
+            - site
+
+  push-adhoc-docs:
+    executor: default_build_environment
+    working_directory: ~/repo
+
+    steps:
+      - checkout
+      - bootstrap_node
+      - bootstrap_aws_cli
+      - attach_workspace:
+          at: ~/repo
+      - run:
+          name: Deploy Adhoc
+          command: 'make publish-docs-adhoc FLAGS=--forceUpload'
+      # TODO: add a probe step here to validate?
+
+  push-head-docs:
+    executor: default_build_environment
+    working_directory: ~/repo
+
+    steps:
+      - checkout
+      - bootstrap_node
+      - bootstrap_aws_cli
+      - attach_workspace:
+          at: ~/repo
+      - run:
+          name: Deploy Head
+          command: 'make publish-docs-head FLAGS=--forceUpload'
+      # TODO: add a probe step here to validate?
+  
+  push-staging-docs:
+    executor: default_build_environment
+    working_directory: ~/repo
+
+    steps:
+      - checkout
+      - bootstrap_node
+      - bootstrap_aws_cli
+      - attach_workspace:
+          at: ~/repo
+      - run:
+          name: Deploy Staging
+          command: 'make publish-docs-staging FLAGS=--forceUpload'
+      # TODO: add a probe step here to validate?
+  
+  push-prod-docs:
+    executor: default_build_environment
+    working_directory: ~/repo
+
+    steps:
+      - checkout
+      - bootstrap_node
+      - bootstrap_aws_cli
+      - attach_workspace:
+          at: ~/repo
+      - run:
+          name: Deploy Prod
+          command: 'make publish-docs-prod FLAGS=--forceUpload'
+      # TODO: add a probe step here to validate?
+
 workflows:
   version: 2
   commit_validation:
     jobs:
       - build:
           context:
             - dockerhub
+
+  deploy:
+    jobs:
+      - build:
+          context:
+            - dockerhub
+          filters:
+            branches:
+              only:
+                - release
+      - utils/cancel-older-awaiting-approvals:
+          context:
+            - dockerhub
+            - util
+          circle_token: ${CIRCLECI_TOKEN}
+          requires:
+            - build
+      - build-docs:
+          context:
+            - dockerhub
+          requires:
+            - build      
+      # TODO(spencer): figure out why we aren't serving CDN on adhoc, technically this step isn't needed
```

Comment:
I am not sure why it's necessary to push to adhoc. or if we meant to push to all envs (including adhoc), what do you mean by we are not serving CDN on adhoc?